### PR TITLE
process attach/detach sample

### PR DESCRIPTION
**Pull request by bgopesh**
## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: projects/rocprofiler-sdk/samples/process_attachment/client.cpp [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+    // Detach from the process
+    if(!client.detach())
+    {
+        std::cerr << "Warning: Failed to cleanly detach from process" << std::endl;
+    }
+
+    std::cout << "Process attachment session completed" << std::endl;
+    return 0;
+}
```
Comment (meserve-amd) [inline projects/rocprofiler-sdk/samples/process_attachment/client.cpp]: Add newline

File: projects/rocprofiler-sdk/samples/process_attachment/client.cpp [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+        
+        // Monitor target process while profiling
+        for(int i = 0; i < duration_seconds; ++i)
+        {
+            if(!process_exists(target_pid))
+            {
+                std::cout << "Target process " << target_pid << " has exited" << std::endl;
+                break;
+            }
+            std::this_thread::sleep_for(std::chrono::seconds(1));
```
Comment (meserve-amd) [inline projects/rocprofiler-sdk/samples/process_attachment/client.cpp]: Sleep is not a robust timing mechanism, consider swapping this loop to use a std::async and std::future
Launch a std::async with a function that repeatedly checks the PID and sleeps
You could also have that function compare to a starting timepoint to keep the progress messages
Use std::future::wait_for to enforce the total duration

File: projects/rocprofiler-sdk/samples/process_attachment/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+    )
+endif()
+
+# Install targets
+install(
+    TARGETS process-attachment-tool process-attachment-client process-attachment-target
+    DESTINATION ${CMAKE_INSTALL_BINDIR}
+    COMPONENT samples
+)
```
Comment (meserve-amd) [inline projects/rocprofiler-sdk/samples/process_attachment/CMakeLists.txt]: Add newline

File: projects/rocprofiler-sdk/samples/process_attachment/main.cpp [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+        std::cout << "Application finished normally" << std::endl;
+    }
+    catch(const std::exception& e)
+    {
+        std::cerr << "Error: " << e.what() << std::endl;
+        return 1;
+    }
+    
+    return 0;
+}
```
Comment (meserve-amd) [inline projects/rocprofiler-sdk/samples/process_attachment/main.cpp]: Add newline

File: projects/rocprofiler-sdk/samples/process_attachment/tool.cpp [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+    attachment_tool::set_kernel_tracing(enable);
+}
+
+void
+rocprofiler_set_hip_api_tracing(bool enable)
+{
+    attachment_tool::set_hip_api_tracing(enable);
+}
+
+}  // extern "C"
```
Comment (meserve-amd) [inline projects/rocprofiler-sdk/samples/process_attachment/tool.cpp]: Add newline

File: projects/rocprofiler-sdk/samples/process_attachment/tool.hpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+void
+set_kernel_tracing(bool enable);
+
+/**
+ * @brief Enable/disable HIP API tracing
+ */
+void
+set_hip_api_tracing(bool enable);
+
+}  // namespace attachment_tool
```
Comment (meserve-amd) [inline projects/rocprofiler-sdk/samples/process_attachment/tool.hpp]: Add newline

File: projects/rocprofiler-sdk/samples/process_attachment/tool.cpp [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+rocprofiler_set_kernel_tracing(bool enable)
+{
+    attachment_tool::set_kernel_tracing(enable);
+}
+
+void
+rocprofiler_set_hip_api_tracing(bool enable)
+{
+    attachment_tool::set_hip_api_tracing(enable);
+}
```
Comment (meserve-amd) [inline projects/rocprofiler-sdk/samples/process_attachment/tool.cpp]: Are these used anywhere? Remove if not.
Drop the rocprofiler_ namespacing if they are used

File: projects/rocprofiler-sdk/samples/process_attachment/tool.cpp [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+// Forward declarations for attach functions
+namespace attachment_tool
+{
+int tool_attach(rocprofiler_client_detach_t detach_func,
+               rocprofiler_context_id_t*   context_ids,
+               uint64_t                    context_ids_length,
+               void*                       tool_data);
+
+void tool_detach(void* tool_data);
+}  // namespace attachment_tool
```
Comment (meserve-amd) [inline projects/rocprofiler-sdk/samples/process_attachment/tool.cpp]: Are these necessary?

File: projects/rocprofiler-sdk/samples/process_attachment/tool.cpp [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+    std::string output_filename = "process_attachment_trace_" + std::to_string(getpid()) + ".log";
+    output_file.open(output_filename, std::ios::out | std::ios::trunc);
+    
+    {
+        std::lock_guard<std::mutex> lock(output_mutex);
+        if(output_file.is_open())
+        {
+            output_file << "=== Process Attachment Tool Started ===" << std::endl;
+            output_file << "PID: " << getpid() << std::endl;
+            output_file << "========================================" << std::endl;
```
Comment (meserve-amd) [inline projects/rocprofiler-sdk/samples/process_attachment/tool.cpp:245]:
```suggestion
            output_file << "=======================================" << std::endl;
```

**Review by meserve-amd — CHANGES_REQUESTED** [review]

File: projects/rocprofiler-sdk/samples/process_attachment/client.cpp (hunk 1 above)
Comment (bgopesh) [reply projects/rocprofiler-sdk/samples/process_attachment/client.cpp]: done

File: projects/rocprofiler-sdk/samples/process_attachment/CMakeLists.txt (hunk 3 above)
Comment (bgopesh) [reply projects/rocprofiler-sdk/samples/process_attachment/CMakeLists.txt]: done

File: projects/rocprofiler-sdk/samples/process_attachment/tool.cpp (hunk 5 above)
Comment (bgopesh) [reply projects/rocprofiler-sdk/samples/process_attachment/tool.cpp]: done

File: projects/rocprofiler-sdk/samples/process_attachment/main.cpp (hunk 4 above)
Comment (bgopesh) [reply projects/rocprofiler-sdk/samples/process_attachment/main.cpp]: done

File: projects/rocprofiler-sdk/samples/process_attachment/client.cpp (hunk 2 above)
Comment (bgopesh) [reply projects/rocprofiler-sdk/samples/process_attachment/client.cpp]: done

File: projects/rocprofiler-sdk/samples/process_attachment/tool.cpp (hunk 7 above)
Comment (bgopesh) [reply projects/rocprofiler-sdk/samples/process_attachment/tool.cpp]: done

File: projects/rocprofiler-sdk/samples/process_attachment/tool.cpp (hunk 8 above)
Comment (bgopesh) [reply projects/rocprofiler-sdk/samples/process_attachment/tool.cpp]: removed

**Review by meserve-amd — APPROVED** [review]
Looks good. Thank you for tackling this.

**Comment by bgopesh** [comment]
This needs to be readdressed in a separate PR as the APIs have changed.
